clippy_lints/src/operators/manual_is_multiple_of.rs

```diff
@@ -0,0 +1,115 @@
+use clippy_utils::consts::{integer_const, is_zero_integer_const};
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::msrvs::{self, Msrv};
+use clippy_utils::source::SpanRangeExt;
+use clippy_utils::sugg::Sugg;
+use rustc_ast::BinOpKind;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty;
+
+use super::MANUAL_IS_MULTIPLE_OF;
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &Expr<'_>,
+    op: BinOpKind,
+    lhs: &'tcx Expr<'tcx>,
+    rhs: &'tcx Expr<'tcx>,
+    min_and_mask_size: u8,
+    msrv: Msrv,
+) {
+    if let Some(operand) = uint_compare_to_zero(cx, op, lhs, rhs)
+        && let ExprKind::Binary(lhs_op, lhs_left, lhs_right) = operand.kind
+        && msrv.meets(cx, msrvs::UNSIGNED_IS_MULTIPLE_OF)
+    {
+        let mut app = Applicability::MachineApplicable;
+        let (dividend, divisor) = if lhs_op.node == BinOpKind::Rem {
+            (
+                lhs_left,
+                Sugg::hir_with_applicability(cx, lhs_right, "_", &mut app).into_string(),
+            )
+        } else if lhs_op.node == BinOpKind::BitAnd {
+            let min_divisor = 1 << u128::from(min_and_mask_size);
+            if let Some(divisor) = is_all_ones(cx, lhs_right, min_divisor, &mut app) {
+                (lhs_left, divisor)
+            } else if let Some(divisor) = is_all_ones(cx, lhs_left, min_divisor, &mut app) {
+                (lhs_right, divisor)
+            } else {
+                return;
+            }
+        } else {
+            return;
+        };
+        span_lint_and_sugg(
+            cx,
+            MANUAL_IS_MULTIPLE_OF,
+            expr.span,
+            "manual implementation of `.is_multiple_of()`",
+            "replace with",
+            format!(
+                "{}{}.is_multiple_of({divisor})",
+                if op == BinOpKind::Eq { "" } else { "!" },
+                Sugg::hir_with_applicability(cx, dividend, "_", &mut app).maybe_par()
+            ),
+            app,
+        );
+    }
+}
+
+// If we have a `x == 0`, `x != 0` or `x > 0` (or the reverted ones), return the non-zero operand
+fn uint_compare_to_zero<'tcx>(
+    cx: &LateContext<'tcx>,
+    op: BinOpKind,
+    lhs: &'tcx Expr<'tcx>,
+    rhs: &'tcx Expr<'tcx>,
+) -> Option<&'tcx Expr<'tcx>> {
+    let operand = if matches!(lhs.kind, ExprKind::Binary(..))
+        && matches!(op, BinOpKind::Eq | BinOpKind::Ne | BinOpKind::Gt)
+        && is_zero_integer_const(cx, rhs)
+    {
+        lhs
+    } else if matches!(rhs.kind, ExprKind::Binary(..))
+        && matches!(op, BinOpKind::Eq | BinOpKind::Ne | BinOpKind::Lt)
+        && is_zero_integer_const(cx, lhs)
+    {
+        rhs
+    } else {
+        return None;
+    };
+
+    matches!(cx.typeck_results().expr_ty_adjusted(operand).kind(), ty::Uint(_)).then_some(operand)
```

Comment:
Are we interested in also accounting for constants?

Reply:
What do you mean? `CONST & MASK == 0` will already be checked if in an unsigned context. Are you suggesting we lint them if both are constants and typed as signed values?

Reply:
(gentle ping to @blyxyas)